modelforge/dataset/utils.py

```diff
@@ -126,6 +126,161 @@ def split(self, dataset: "TorchDataset") -> Tuple[Subset, Subset, Subset]:
+
+        Parameters
+        ----------
+        dataset : TorchDataset
+            The dataset to be split.
+
+        Returns
+        -------
+        Tuple[Subset, Subset, Subset]
+            A tuple containing three Subsets for training, validation, and testing subsets, respectively.
+
+        Examples
+        --------
+        >>> dataset = TorchDataset(numpy_data)
+        >>> random_strategy = RandomSplittingStrategy(seed=42, split=[0.7, 0.2, 0.1])
+        >>> train_dataset, val_dataset, test_dataset = random_strategy.split(dataset)
+        """
+
+        logger.debug(f"Using random splitting strategy with seed {self.seed} ...")
+        logger.debug(
+            f"Splitting dataset into {self.train_size}, {self.val_size}, {self.test_size} ..."
+        )
+
+        train_d, val_d, test_d = random_record_split(
+            dataset,
+            lengths=[self.train_size, self.val_size, self.test_size],
+            generator=self.generator,
+        )
+
+        return (train_d, val_d, test_d)
+
+
+def random_record_split(
+        dataset: "TorchDataset",
+        lengths: List[Union[int, float]],
+        generator: Optional[torch.Generator] = torch.default_generator
+) -> List[Subset]:
+    """
+    Randomly split a TorchDataset into non-overlapping new datasets of given lengths, keeping all conformers in a record in the same split
+
+    If a list of fractions that sum up to 1 is given,
+    the lengths will be computed automatically as
+    floor(frac * len(dataset)) for each fraction provided.
+
+    After computing the lengths, if there are any remainders, 1 count will be
+    distributed in round-robin fashion to the lengths
+    until there are no remainders left.
+
+
+    Parameters
+    ----------
+    dataset : TorchDataset
+        Dataset to be split.
+    lengths : List[int]
+        Lengths of splits to be produced.
+    generator : Optional[torch.Generator], optional
+        Generator used for the random permutation, by default None
+
+    Returns
+    -------
+    List[Subset]
+        List of subsets of the dataset.
+
+    """
+    if np.isclose(sum(lengths), 1) and sum(lengths) <= 1:
+        subset_lengths: List[int] = []
+
+        for i, frac in enumerate(lengths):
+            if frac < 0 or frac > 1:
+                raise ValueError(f"Fraction at index {i} is not between 0 and 1")
+            n_items_in_split = int(
+                np.floor(dataset.record_len() * frac)  # type: ignore[arg-type]
+            )
+            subset_lengths.append(n_items_in_split)
+
+        remainder = dataset.record_len() - sum(subset_lengths)  # type: ignore[arg-type]
+
+        # add 1 to all the lengths in round-robin fashion until the remainder is 0
+        for i in range(remainder):
+            idx_to_add_at = i % len(subset_lengths)
+            subset_lengths[idx_to_add_at] += 1
+
+        lengths = subset_lengths
+
+        for i, length in enumerate(lengths):
+            if length == 0:
```

Comment:
this results in an empty dataset, can this be intended behavior?

Reply:
Perhaps the user intends to create an empty dataset, but this will warn the user that they are doing so. I copied this from the implementation of `torch.utils.data.random_split`

Reply:
Thanks for clarifying